src/librustc_metadata/rmeta/decoder.rs

```diff
@@ -485,23 +492,9 @@ impl<'a, 'tcx> Decodable<DecodeContext<'a, 'tcx>> for Span {
         // treat the 'local' and 'foreign' cases almost identically during deserialization:
         // we can call `imported_source_files` for the proper crate, and binary search
         // through the returned slice using our span.
-        let imported_source_files = if tag == TAG_VALID_SPAN_LOCAL {
+        let imported_source_files = if cnum == LOCAL_CRATE {
```

Comment:
You're right - this is causing us to call `imported_source_files` with a bunch of extra indirection (going through the `CrateStore` instead of just using our local `decoder.cdata()`. Additionally, this will cause `last_source_file_index` to never be used - hopefully that's responsible for the slowdown.